backend/canvas_app_explorer/lti1p3.py

```diff
@@ -16,20 +17,34 @@
 from pylti1p3.contrib.django import DjangoOIDCLogin, DjangoMessageLaunch, \
     DjangoCacheDataStorage, DjangoDbToolConf
 
+
 logger = logging.getLogger(__name__)
 
+
+class CanvasRole(Enum):
+    ACCOUNT_ADMIN = 'Account Admin'
+    SUB_ACCOUNT_ADMIN = 'Sub-Account Admin'
+    TEACHER = 'TeacherEnrollment'
+    STUDENT = 'StudentEnrollment'
```

Comment:
Maybe no need to define the student role if student is not allowed to use the tool

Reply:
Also these roles are already defined in the library. I'm not sure we need to redefine them.

https://github.com/dmitry-viskov/pylti1.3#check-users-role-after-lti-launch

We should just be able to check this with

`message_launch.check_staff_access() or message_launch.check_admin_access()`

It doesn't look like there's anything specific about sub account admin but we don't need that yet role yet either.

Reply:
This is also what CCM does.

Reply:
I see, so there has been some discussion lately about [supporting sending a masquerade value](https://github.com/IMSGlobal/LTI-central/issues/157) over in the LTI launch. Right now I think you'd have to send that as a custom parameter if you cared about it. (Like we might if/when we supported students)

I'd believe this is an edge case we don't have to be concerned with now. If they're masquerading as a student the instructor won't have access to the tool in the navigation after the change here. And it's not going to give them a more limited scoped token if they did access it directly. 

I'm also not sure why it wouldn't be working.

Launching as a Instructor/Admin I see
```
  "https://purl.imsglobal.org/spec/lti/claim/roles": [
    "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Administrator",
    "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Instructor",
    "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Student",
    "http://purl.imsglobal.org/vocab/lis/v2/membership#Instructor",
    "http://purl.imsglobal.org/vocab/lis/v2/system/person#User"
  ],
```
and
```
  "https://purl.imsglobal.org/spec/lti/claim/custom": {
    "roles": "TeacherEnrollment,Account Admin",
```

Doing Student View it has
```
  "https://purl.imsglobal.org/spec/lti/claim/roles": [
    "http://purl.imsglobal.org/vocab/lis/v2/institution/person#Student",
    "http://purl.imsglobal.org/vocab/lis/v2/membership#Learner",
    "http://purl.imsglobal.org/vocab/lis/v2/system/person#User",
    "http://purl.imsglobal.org/vocab/lti/system/person#TestUser"
  ],
```
and

```
  "https://purl.imsglobal.org/spec/lti/claim/custom": {
    "roles": "StudentEnrollment",
```

Reply:
I think it's fine since we're a little time crunched to go with this approach you have here for now. But we should file an issue for looking to support the standard LTI Roles when/if we support more roles and features in the future.

Reply:
Ah, I misread this as Student View verses Masquerade. This action in our instance is limited to someone that's an admin (or account admin).

It looks like it still sends `Instructor` as an institution role and the pylti1.3 [includes this as a common role](https://github.com/dmitry-viskov/pylti1.3/blob/master/pylti1p3/roles.py) for a Teacher. 

I'd think if you wanted a non masqueraded common role here you'd have to have only a context role. Though I don't think this "View As" edge case is going to really be a problem. They still do technically have the role as an Instructor and a Student in the same course. 

```
class TeacherRoleContextOnly(AbstractRole):
    _context_roles = ('Instructor', 'Administrator')  # type: tuple
```

Reply:
I tried something like this class, and it's not really working as I expected. Code below:

```
class StaffRoleContextOnly(AbstractRole):
    _institution_roles = ('Administrator',)
    _context_roles = ('Instructor', 'Administrator')
```

```
    logger.info(roles)
    staff_role_obj = StaffRoleContextOnly(message_launch._get_jwt_body())
    for role in roles:
        logger.info(staff_role_obj.parse_role_str(role))

    user_is_staff_in_course = StaffRoleContextOnly(message_launch._get_jwt_body()).check()
    
    if not user_is_staff_in_course:
        logger.warn(f'User {username} does not have a staff role.')
        raise PermissionDenied()
```

This will work for the admin and student cases, but I couldn't launch when I was masquerading as an instructor. My initial impression is that the context roles generated by Canvas are malformed, or there's a bug in this library, but not sure exactly. When I parse the strings Canvas generates as the library does, I get this:
```
[INFO] [lti1p3.py:145] ['http://purl.imsglobal.org/vocab/lis/v2/institution/person#Instructor', 'http://purl.imsglobal.org/vocab/lis/v2/institution/person#Student', 'http://purl.imsglobal.org/vocab/lis/v2/membership#Instructor', 'http://purl.imsglobal.org/vocab/lis/v2/system/person#User']
canvas_app_explorer        | [2022-07-20 17:32:33 +0000] canvas_app_explorer        | [2022-07-20 17:32:33 +0000] [INFO] [lti1p3.py:148] ('Student', 'institution')
[2022-07-20 17:32:33 +0000] [INFO] [lti1p3.py:148] ('Instructor', None)
[2022-07-20 17:32:33 +0000] [INFO] [lti1p3.py:148] ('User', 'system')

Reply:
@jonespm, it looks there may be an issue with the `pylti1.3` source code:
https://github.com/dmitry-viskov/pylti1.3/blob/7474685aecc35d528b944caf05b28ea66ac6d204/pylti1p3/roles.py#L44-L57

http://www.imsglobal.org/spec/lti/v1p3/#lis-vocabulary-for-context-roles

The `membership` roles just don't have the same structure as the institution or system roles; there is no slash, so the splits don't result in `('Instructor`, 'membership')` as expected.

I think I'm going to stick with the custom variable roles approach for now, but sure, we can create an issue to do this. It'd be interesting to look into this for CCM as well, but I don't think `ltijs` has an equivalent role class. Then again, these tools are pretty reliant on Canvas, and we're not getting much from the LMS-agnostic aspect of the LTI/LIS roles.

Reply:
@zqian , I removed the Student role. We may need it in the future. https://github.com/tl-its-umich-edu/canvas-app-explorer/pull/247/commits/76a6c58050af25aba64d2af7ecd4b47a92844902

Reply:
I'm fine with making a new issue and looking into it. But I'm also not convinced the "Act As" is really a true use case either. I think we'll have to pass the `Canvas.masqueradingUser.id` as a custom parameter until a better solution is worked out by IMS if we really cared about this use case. 

It does say when you "Act As" that 

> "You will be able to take any action as if you were this user, and from other users' points of views, it will be as if this user performed them. However, audit logs record that you were the one who performed the actions on behalf of this user."

So you still are the kind of both users when doing this. 

When the author committed this he said that pull requests were welcome against this if it's not splitting it correctly. I was thinking it probably could use some unit tests but wasn't about to prioritize that right now.

Reply:
Yeah, probably a pull request with unit tests would be a good addition if we wanted to rely on this.

I'm still not quite sure what you mean by "I'm also not convinced the "Act As" is really a true use case either". The role identifiers I listed above for a masqueraded instructor don't show admin roles, and at least as far as the context ('`membership`, what's at issue here) there's no indication someone is masquerading as the user.

Reply:
I will create a new issue.